<?xml version="1.0" ?>
<!-- [6-11-24] Constructed for simulating a 3 DoF SMM, with 3 joint effort controllers && R-idosc centralized node -->
<!-- [26-12-24] Resumed working. It implements the algorithm in MATLAB script smm_ridosc_test2.m-->
     
<launch>
    <!-- Define folder name for rosbags -->
    <arg name="run_name" default="hridosc_default_run" />
    <!-- Define robot specific arguments -->
    <arg name="tcp_link_name" default="massage_tool" />
    <arg name="str_digit_glob" default="3" />

    <!-- Define global parameters-->
    <param name="str_digit" value="$(arg str_digit_glob)" />
    <param name="base_frame" value="base_plate" />
    <param name="tcp_frame" value="massage_tool" />

    <!-- Include YAML file with RIDOSC parameters -->
    <rosparam file="$(find smm_control)/config/yaml/ridosc/fixed_params_full.yaml" command="load"/>

    <!-- Include launch file Gazebo with robot and all controller settings -->
    <include file="$(find smm_control)/launch/idosc_gazebo_settings.launch"/> <!--use same settings for idosc/fasmc/ridosc -->
    
    <!-- R-idosc SIMPLE IMPLEMENTATION --> 
    <!-- I. PUBLISH DESIRED STATES -->
    <!-- Node that publishes to /tcp_desired_state -->
    <rosparam file="$(find smm_control)/config/yaml/paths/tspace_goal.yaml" command="load"/>
    <node name="publishSingleTcpGoal" pkg="smm_control" type="publishSingleTcpGoal" output="screen">
        <param name="TCP_LINK_NAME" value="$(arg tcp_link_name)" />     
    </node>
    <!-- Node that handles the mismatch between tcp goal and ridosc desired state msg types -->
    <!-- Node that publishes to /ridosc_desired_state -->
    <node name="updateDesiredState_ridosc_simple" pkg="smm_control" type="updateDesiredState_ridosc_simple" output="screen" />

    <!-- II. ROBOT DATA -->
    <!-- Node that generates the topic /joint_current_state -->
    <node name="updateJointState_simple" pkg="smm_control" type="updateJointState_simple" output="screen" />

    <!-- Analytical Dynamics using screws @ /joint_current_state -->
    <node name="serverOperationalSpaceMatrices" pkg="smm_control" type="serverOperationalSpaceMatrices" output="screen"/>

    <!-- Analytical Jacobians using screws @ /joint_current_state -->
    <!--<node name="serverToolJacobians" pkg="smm_control" type="serverToolJacobians" output="screen" />-->

    <!-- Service that computes TCP velocity and acceleration numerically from tf position data-->
    <!--<node name="serverTcpStateTfNum" pkg="smm_control" type="serverTcpStateTfNum" output="screen"/>-->

    <!-- III. RIDOSC -->
    <!-- Node that generates the topic /tcp_current_state -->
    <node name="updateCartesianState_ridosc_simple" pkg="smm_control" type="updateCartesianState_ridosc_simple" output="screen"/>

    <!-- Node that publishes to /ridosc_error_state -->
    <node name="updateErrorState_ridosc_simple" pkg="smm_control" type="updateErrorState_ridosc_simple" output="screen" />
    
    <!-- Sliding state -->
    <node name="updateSlidingState_ridosc_simple" pkg="smm_control" type="updateSlidingState_ridosc_simple" output="screen" />

    <!-- Fuzzy Parameters: Not used in ridosc! -->
    <!--<node name="updateFuzzyParams_fasmc_simple" pkg="smm_control" type="updateFuzzyParams_fasmc_simple" output="screen" /> -->

    <!-- Adaptive Parameters -->
    <node name="updateParams_ridosc_simple" pkg="smm_control" type="updateParams_ridosc_simple" output="screen" />

    <!-- Dynamic Model torque term -->  
    <node name="updateDynamicsTorque_ridosc_simple" pkg="smm_control" type="updateDynamicsTorque_ridosc_simple" output="screen" />

    <!-- Robust torque term -->
    <node name="updateRobustTorque_ridosc_simple" pkg="smm_control" type="updateRobustTorque_ridosc_simple" output="screen" />

    <!-- Add ridosc torques-->
    <node name="updateTorqueCommand_ridosc_simple" pkg="smm_control" type="updateTorqueCommand_ridosc_simple" output="screen" /> 

    <!-- Publish to robot effort controllers -->
    <node name="robot_torque_command_publisher" pkg="smm_control" type="robot_torque_command_publisher" output="screen" /> 

    <!-- IV. DATA LOGGING -->
    <!-- Include launch file for data logging and real time monitoring -->
    <include file="$(find smm_control)/launch/ridosc_data_monitor.launch">
        <arg name="run_name" value="$(arg run_name)"/>
    </include>

    <!-- V. RviZ Debugger-->
    <include file="$(find smm_control)/launch/ridosc_visualization_simple.launch"/> <!--use same settings for idosc/fasmc/ridosc -->
</launch>